### PR TITLE
Stop streams on finish

### LINKED ISSUE #263: Poison streams after the app func completes
Any attempt to write to the response body or read from the request body after the app func completes should fail with an `ObjectDisposedException`.

### PR DESCRIPTION
**Pull request by benaadams**
Can't use dispose (or close) as can be disposed too early by user code

Resolves #263

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs [hunk 1]
```diff
@@ -102,5 +123,10 @@ public override void Write(byte[] buffer, int offset, int count)
         {
             throw new NotImplementedException();
         }
+
+        public void StopAcceptingReads()
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: It's worth including your comment here about why you can't use Dispose.

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: Added and referenced example of this

File: src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs [hunk 2]
```diff
@@ -50,12 +51,22 @@ public override void SetLength(long value)
 
         public override int Read(byte[] buffer, int offset, int count)
         {
+            if (_stopped)
+            {
+                throw new ObjectDisposedException("RequestStream has been disposed");
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: This ctor expects the object name instead of an exception message.

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/FrameRequestStream.cs]: Corrected

**Comment by benaadams** [comment]
Bumped for CI retest rather than doing a whitespace change and force push as branch https://github.com/aspnet/KestrelHttpServer/tree/shalter/request-lifetime has built on it.

**Comment by NickCraver** [comment]
+1

**Comment by halter73** [comment]
I went against the normal pattern of merging this and rebased instead. It was more convenient since I was working against these changes in my request-lifetime PR.

Here is the last of the 3 relevant commits: https://github.com/aspnet/KestrelHttpServer/commit/1cbb273b0d6d752ed557a78de38a5b83a8223d0e

**Comment by benaadams** [comment]
Yar; you stole my rebase; force push ability ;-)

**Pull request closed without merging**